moto/sns/responses.py

```diff
@@ -459,6 +467,126 @@ def set_subscription_attributes(self):
+
+        if len(filter_list) > 0:
+            result = {k: v for k, v in self.backend.sms_attributes.items() if k in filter_list}
+        else:
+            result = self.backend.sms_attributes
+
+        template = self.response_template(GET_SMS_ATTRIBUTES_TEMPLATE)
+        return template.render(attributes=result)
+
+    def check_if_phone_number_is_opted_out(self):
+        number = self._get_param('phoneNumber')
+        if self.OPT_OUT_PHONE_NUMBER_REGEX.match(number) is None:
+            error_response = self._error(
+                code='InvalidParameter',
+                message='Invalid parameter: PhoneNumber Reason: input incorrectly formatted'
+            )
+            return error_response, dict(status=400)
+
+        # There should be a nicer way to set if a nubmer has opted out
+        template = self.response_template(CHECK_IF_OPTED_OUT_TEMPLATE)
+        return template.render(opt_out=str(number.endswith('99')).lower())
+
+    def list_phone_numbers_opted_out(self):
+        template = self.response_template(LIST_OPTOUT_TEMPLATE)
+        return template.render(opt_outs=self.backend.opt_out_numbers)
+
+    def opt_in_phone_number(self):
+        number = self._get_param('phoneNumber')
+
+        try:
+            self.backend.opt_out_numbers.remove(number)
+        except ValueError:
+            pass
+
+        template = self.response_template(OPT_IN_NUMBER_TEMPLATE)
+        return template.render()
+
+    def add_permission(self):
+        arn = self._get_param('TopicArn')
+        label = self._get_param('Label')
+        accounts = self._get_multi_param('AWSAccountId.member.')
+        action = self._get_multi_param('ActionName.member.')
+
+        if arn not in self.backend.topics:
+            error_response = self._error('NotFound', 'Topic does not exist')
+            return error_response, dict(status=404)
+
+        key = (arn, label)
+        self.backend.permissions[key] = {'accounts': accounts, 'action': action}
+
+        template = self.response_template(ADD_PERMISSION_TEMPLATE)
+        return template.render()
+
+    def remove_permission(self):
+        arn = self._get_param('TopicArn')
+        label = self._get_param('Label')
+
+        if arn not in self.backend.topics:
+            error_response = self._error('NotFound', 'Topic does not exist')
+            return error_response, dict(status=404)
+
+        try:
+            key = (arn, label)
+            del self.backend.permissions[key]
+        except KeyError:
+            pass
+
+        template = self.response_template(DEL_PERMISSION_TEMPLATE)
+        return template.render()
+
+    def confirm_subscription(self):
+        arn = self._get_param('TopicArn')
+
+        if arn not in self.backend.topics:
+            error_response = self._error('NotFound', 'Topic does not exist')
+            return error_response, dict(status=404)
+
+        # Added other parts here for when they are needed
+        # token = self._get_param('Token')
```

Comment:
@terrycain Is this meant as a TODO for implementing features later? If so, a developer stumbling on this could use some direction as to what's missing that prevents this from being actual code so they know what to implement. Otherwise maybe just have a reference to what feature needs to be implemented?

Reply:
Yeah basically, when you subscribe an entity to an SNS topic, you'll get an out-of-band token, be it an automated email, or an sms with a code. This is then entered to say, "yes I consent to being subscribed to the topic.". We currently don't do anything to 1. generate the token and store it, and 2. offer a mechanism to asyncronously dish out the token on whatever contact method they chose.

It was meant to be a partial implementation for when/if we validate tokens. Will update L577 to be a bit more descriptive.